File: flutter_matter_android/android/src/main/kotlin/net/grandcentrix/flutter_matter/clusters/OnOffCluster.kt
package net.grandcentrix.flutter_matter.clusters

import chip.devicecontroller.ChipClusters
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import net.grandcentrix.flutter_matter.AndroidError
import net.grandcentrix.flutter_matter.FlutterError
import net.grandcentrix.flutter_matter.FlutterMatterFlutterOnOffClusterApi
import net.grandcentrix.flutter_matter.FlutterMatterHostOnOffClusterApi
import net.grandcentrix.flutter_matter.chip.ChipClient
import timber.log.Timber
import java.io.Closeable

class OnOffCluster(private val flutterOnOffClusterApi: FlutterMatterFlutterOnOffClusterApi) : FlutterMatterHostOnOffClusterApi, Closeable {
    private lateinit var chipClient: ChipClient

    var activity: android.app.Activity? = null
        set(value) {
            if (value == activity) return
            field = value
            if (value != null) {
                chipClient = ChipClient(value!!)
            }
        }

    private val scope =
        CoroutineScope(Dispatchers.IO + Job())

    private suspend fun getCluster(
        deviceId: Long,
        endpointId: Long,
    ): ChipClusters.OnOffCluster {
        val devicePtr = chipClient.getConnectedDevicePointer(deviceId)
        return ChipClusters.OnOffCluster(devicePtr, endpointId.toInt())
    }

    override fun off(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Unit>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster
                ?.off(
                    object : ChipClusters.DefaultClusterCallback {
                        override fun onSuccess() {
                            callback(Result.success(Unit))
                        }

                        override fun onError(ex: Exception) {
                            Timber.e(ex, "Failed to send command off")
                            callback(
                                Result.failure(
                                    FlutterError(
                                        "-1",
                                        "Failed to send command off",
                                    ),
                                ),
                            )
                        }
                    },
                )
        }
    }

    override fun on(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Unit>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster
                ?.on(
                    object : ChipClusters.DefaultClusterCallback {
                        override fun onSuccess() {
                            callback(Result.success(Unit))
                        }

                        override fun onError(ex: Exception) {
                            Timber.e(ex, "Failed to send command on")
                            callback(
                                Result.failure(
                                    FlutterError(
                                        "-1",
                                        "Failed to send command on",
                                    ),
                                ),
                            )
                        }
                    },
                )
        }
    }

    override fun toggle(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Unit>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster
                ?.toggle(
                    object : ChipClusters.DefaultClusterCallback {
                        override fun onSuccess() {
                            callback(Result.success(Unit))
                        }

                        override fun onError(ex: Exception) {
                            Timber.e(ex, "Failed to send command toggle")
                            callback(
                                Result.failure(
                                    FlutterError(
                                        "-1",
                                        "Failed to send command toggle",
                                    ),
                                ),
                            )
                        }
                    },
                )
        }
    }

    override fun readOnOff(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster
                ?.readOnOffAttribute(
                    object : ChipClusters.BooleanAttributeCallback {
                        override fun onSuccess(value: Boolean) {
                            callback(Result.success(value))
                        }

                        override fun onError(ex: Exception) {
                            Timber.e(ex, "Failed to read attribute OnOff")
                            callback(
                                Result.failure(
                                    FlutterError(
                                        "-1",
                                        "Failed to read attribute OnOff",
                                    ),
                                ),
                            )
                        }
                    },
                )
        }
    }

    override fun subscribeToOnOff(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Unit>) -> Unit,
    ) {
        scope.launch {
            val cluster =
                try {
                    getCluster(deviceId, endpointId)
                } catch (e: Exception) {
                    callback(Result.failure(FlutterError("-1", "Can't connect to device!")))
                    null
                }

            cluster?.subscribeOnOffAttribute(
                object : ChipClusters.BooleanAttributeCallback {
                    override fun onSuccess(value: Boolean) {
                        activity!!.runOnUiThread {
                            flutterOnOffClusterApi.onOff(deviceId, endpointId, value, null) {
                                Timber.i("Reported OnOff status: $value")
                            }
                        }
                    }

                    override fun onError(ex: Exception) {
                        activity!!.runOnUiThread {
                            flutterOnOffClusterApi.onOff(
                                deviceId,
                                endpointId,
                                null,
                                AndroidError("-1", ex.message),
                            ) {
                                Timber.e(ex, "Reported OnOff error")
                            }
                        }
                    }
                },
                1,
                10,
            )

            if (cluster != null) {
                callback(Result.success(Unit))
            }
        }
    }

    override fun unsubscribeToOnOff(
        deviceId: Long,
        endpointId: Long,
        callback: (Result<Unit>) -> Unit,
    ) {
        // TODO: How? ¯\_(ツ)_/¯
        callback(Result.success(Unit))
    }

    override fun close() {
        scope.cancel()
    }
}

File: flutter_matter_ios/ios/Classes/Clusters/OnOffCluster.swift
//
//  OnOffCommandHandler.swift
//  flutter_matter_ios
//
//  Created by Philipp Manstein on 19.09.23.
//

import Foundation
import OSLog
import Matter
import Flutter

class OnOffCluster: FlutterMatterHostOnOffClusterApi {
    
    private var _flutterApi: FlutterMatterFlutterOnOffClusterApi;

    init(flutterApi: FlutterMatterFlutterOnOffClusterApi) {
        _flutterApi = flutterApi;
    }
    
    private func getCluster(deviceId: Int64, endpointId: Int64) throws -> MTRBaseClusterOnOff {
        let controller = try MTRDeviceController.shared()
        let device = MTRBaseDevice(nodeID: NSNumber(value: deviceId), controller: controller)
        return MTRBaseClusterOnOff(device: device, endpointID: NSNumber(value: endpointId), queue: DispatchQueue.main)!
    }
    
    // MARK: Commands
    
    func off(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                try await cluster.off();
                completion(Result.success(Void()))
            }
            catch {
                os_log(.error, "Failed to send command off: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to send command off", details: nil)))
            }
        }
    }
    
    func on(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Void, Error>) -> Void)  {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                try await cluster.on();
                completion(Result.success(Void()))
            }
            catch {
                os_log(.error, "Failed to send command on: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to send command on", details: nil)))
            }
        }
    }
    
    func toggle(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Void, Error>) -> Void)  {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                try await cluster.toggle();
                completion(Result.success(Void()))
            }
            catch {
                os_log(.error, "Failed to send command toggle: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to send command toggle", details: nil)))
            }
        }
    }
    
    // MARK: Attributes
    
    func readOnOff(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Bool, Error>) -> Void)  {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let result = try await cluster.readAttributeOnOff();
                completion(Result.success(result == 1 ? true : false))
            }
            catch {
                os_log(.error, "Failed to read attribute OnOff: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute OnOff", details: nil)))
            }
        }
    }
    
    func subscribeToOnOff(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                let cluster = try getCluster(deviceId: deviceId, endpointId: endpointId)
                let params: MTRSubscribeParams = MTRSubscribeParams(minInterval: 1, maxInterval: 10)
                var established = false
                cluster.subscribeAttributeOnOff(with: params,
                                                subscriptionEstablished: {
                    if(established)
                    {
                        return
                    }
                    
                    completion(Result.success(Void()))
                    established = true
                    
                },
                                                reportHandler: { status, error in
                    if(error != nil)
                    {
                        os_log(.error, "Failed report attribute OnOff: \(error)")
                        self._flutterApi.onOff(deviceId: deviceId, endpointId: endpointId, onOff: nil, error: IosError(code: "-1", message: error.debugDescription)) {
                            os_log(.error, "Reported OnOff error: \(error)")
                        }
                        return
                    }
                    
                    self._flutterApi.onOff(deviceId: deviceId, endpointId: endpointId, onOff: status == 1, error: nil) {
                        os_log(.default, "Reported OnOff status: \(status)")
                    }
                });
                
            }
            catch {
                os_log(.error, "Failed to read attribute OnOff: \(error)")
                completion(Result.failure(FlutterError(code: "-1", message: "Failed to read attribute OnOff", details: nil)))
            }
        }
    }
    
    func unsubscribeToOnOff(deviceId: Int64, endpointId: Int64, completion: @escaping (Result<Void, Error>) -> Void) {
        //TODO: How? ¯\_(ツ)_/¯
        completion(Result.success(Void()))
    }
}
